src/rust/src/wallet.rs

```diff
@@ -215,27 +395,43 @@ impl Wallet {
             }
         }
 
-        let tx_notes = self.wallet_tx_notes.get(txid);
+        self.last_observed = Some(LastObserved {
+            block_height,
+            block_tx_idx: Some(block_tx_idx),
+        });
+
+        // update the block height recorded for the transaction
+        if let Some(tx_notes) = self.wallet_received_notes.get_mut(txid) {
+            tx_notes.tx_height = Some(block_height);
+        }
+
+        let my_notes_for_tx = self.wallet_received_notes.get(txid);
         for (action_idx, action) in bundle.actions().iter().enumerate() {
+            // append the note commitment for each action to the witness tree
             if !self
                 .witness_tree
                 .append(&MerkleHashOrchard::from_cmx(action.cmx()))
             {
                 return Err(WalletError::NoteCommitmentTreeFull);
             }
 
-            if let Some(tx_notes) = tx_notes {
-                if tx_notes.decrypted_notes.contains_key(&action_idx) {
-                    self.witness_tree.witness();
-                }
+            // for notes that are ours, witness the current state of the tree
+            if my_notes_for_tx
+                .iter()
+                .any(|n| n.decrypted_notes.contains_key(&action_idx))
```

Comment:
It's really silly to me that Option doesn't have `any` and `all` methods.

Reply:
Not being used to seeing `any`/`all` on `Option`s, I prefer the `map_or` approach since it's more explicit about the default. I think @str4d had a similar comment when we paired on this.

Reply:
I think this is really more of a familiarity thing than anything; existential and universal quantification are absolutely everywhere, and they make just as much sense for Option as for any other container type; in fact, I think that regularly using `exists` and `forall` operations on Option in Scala and on Maybe in Haskell had the side effect of improving my thinking about existential and universal quantification at the type level (where the meaning is closer to that for `Option` than for `List`).